sdk/cosmos/src/clients/cosmos_client.rs

```diff
@@ -134,7 +134,16 @@ impl CosmosClient {
 
     /// Set the auth token used
     pub fn auth_token(&mut self, auth_token: AuthorizationToken) {
-        self.auth_token = auth_token;
+        // TODO: To remove once everything uses the AutorizationPolicy
+        self.auth_token = auth_token.clone();
+
+        // we replace the AuthorizationPolicy. This is
+        // the last-1 policy by construction.
```

Comment:
This seems pretty error prone... I wonder if there's a better way to do it... Perhaps something like this could be better:

```rust
  /// Set the auth token used
    pub fn auth_token(&mut self, auth_token: AuthorizationToken) {
        self.auth_token = auth_token.clone();

        let new_auth_policy: Arc<dyn azure_core::Policy<CosmosContext>> =
            Arc::new(crate::AuthorizationPolicy::new(auth_token));
        use std::any::Any;
        let auth_policy = self
            .pipeline_mut()
            .policies_mut()
            .iter_mut()
            .find(|p| (**p).type_id() == TypeId::of::<AuthorizationPolicy>())
            .unwrap(); // TODO: add auth policy if one doesn't exist?
        std::mem::replace(auth_policy, new_auth_policy);
    }
```
I've not tested whether this works, but I think it should.

Reply:
Yes it is error prone, unfortunately we forego the type safe approach so these problems are bound to happen. 
In this instance, I think we should get away with the simpler code since:

1. The only place we construct the pipeline is `new_pipeline_from_options`.
2. In `new_pipeline_from_options`, we explicitly add the `AuthorizationPolicy` as last retry policy (penultimate position).

Since there is no way to construct a pipeline without an `AuthorizationPolicy` at the penultimate position my prosed code will always work.

At the other hand, your code is safer but it would correctly work on a buggy pipeline (`AuthorizationPolicy` must be the last retry policy for it to work properly) so I think it's a moot point.

What about an assert? Maybe a `debug_assert` that gets removed from production code? 
The assert should enforce that the swapped policy is indeed an `AuthorizationPolicy`.